addons/godot-next/gui/debug_label.gd

```diff
@@ -0,0 +1,155 @@
+# usage:
+#     var debug_label = DebugLabel.new(node)
+#     debug_label.watch(["position:x", "scale", "rotation"])
+#
+# todo: Use RichTextLabel or custom drawing for color coding of core data types.
+#       Unfortunately, it doesn't compute its minimum size the same way as a Label
+
+# Assign a node via inspector. If empty, a parent node is inspected instead.
+export var target_path := NodePath() setget set_target_path
+
+# Inspected object, not restricted to "Node" type, may be assigned via code.
+var target: Object
+
+export var show_label_name := false
+export var show_target_name := false
+
+# A list of property names to be inspected and printed in the `target`.
+# Use indexed syntax (:) to access nested properties: "position:x".
+# Indexing will also work with any `onready` vars defined via script.
+# These can be set beforehand via inspector or via code with "watch()".
+export var properties := PoolStringArray()
+
+# Advanced: depending on the properties inspected, you may need to switch
+# to either "IDLE" or "PHYSICS" update mode to avoid thread issues.
+enum UpdateMode {
+	IDLE,
+	PHYSICS,
+	MANUAL
+}
+export(UpdateMode) var update_mode = UpdateMode.IDLE setget set_update_mode
+
+
+func _init(p_target: Object = null) -> void:
+	if p_target != null:
+		target = p_target
+
+# 	# TODO: RichTextLabel relevant overrides
+#	rect_clip_content = false
+#	scroll_active = false
+#	selection_enabled = true
+
+
+func _enter_tree() -> void:
+	set_process_internal(true)
+
+	if not OS.is_debug_build():
+		text = ""
+		hide()
+		return
+	if target == null:
+		if not target_path.is_empty():
+			_update_target_from_path()
+		else:
+			target = get_parent()
+
+
+func _exit_tree() -> void:
+	set_process_internal(false)
+	set_physics_process_internal(false)
+
+	target = null
+
+
+func set_target_path(p_path: NodePath) -> void:
+	target_path = p_path
+	call_deferred("_update_target_from_path")
+
+
+func set_update_mode(p_mode: int) -> void:
+	update_mode = p_mode
+
+	match update_mode:
+		UpdateMode.IDLE:
+			set_process_internal(true)
+			set_physics_process_internal(false)
+		UpdateMode.PHYSICS:
+			set_process_internal(false)
+			set_physics_process_internal(true)
+		UpdateMode.MANUAL:
+			set_process_internal(false)
+			set_physics_process_internal(false)
+
+
+func _notification(what: int) -> void:
+	# Using internal processing as it guarantees that
+	# debug info is updated even if the scene tree is paused.
+	match what:
+		NOTIFICATION_INTERNAL_PROCESS:
+			_update_debug_info()
+		NOTIFICATION_INTERNAL_PHYSICS_PROCESS:
+			_update_debug_info()
+
+
+func _update_target_from_path() -> void:
+	if has_node(target_path):
+		target = get_node(target_path)
+	# target = get_node_or_null(target_path) # 3.2
+
+
+func watch(p_what) -> void:
+	if p_what is String:
+		var property = p_what
+		properties.push_back(property)
+
+	elif p_what is Array or p_what is PoolStringArray:
+		var list = p_what
+		properties = PoolStringArray(list)
```

Comment:
This sequence is a little inconsistent. If it's a single string, the value gets appended to the list of watched properties. But if it's an array of strings, then it will completely replace the list of properties. How about we make separate methods that handle the appropriate cases in a typed manner?

    func watch(p_what: String):
        properties = PoolStringArray([p_what])
    func watchv(p_what: PoolStringArray):
        properties = p_what
    func watch_append(p_what: String):
        properties.append(p_what)
    func watch_appendv(p_what: PoolStringArray):
        properties.append_array(p_what)

Reply:
That's the beauty of dynamic typing, isn't it? I don't like how static typing trend makes GDScript more difficult to use. Agree with inconsistency but again having 4 separate methods doesn't promote faster prototyping. I'm clearly being subjective now though. 😛

Updated PR with the suggested changes.